### ISSUE TITLE
AzureRM registers for Resource Providers unnecessarily

### ISSUE THREAD

**Issue by colemickens** [hashicorp/terraform#10931]
When replying to this issue (https://github.com/Azure/azure-rest-api-specs/issues/836), I remembered an issue with how AzureRM operates.

**Scenario:**

1. I (an Admin on the Subscription) pre-create resource group `blah`.

2. I create an SP and grant it Contributor level access to `blah`.

3. I run Terraform with those credentials, only asking it to create resources in the resource group that the SP has access to.

This fails today, because every call tries to register the Resource Provider, even if it's already registered and this registration action is at the Subscription scope.

**Suggested Fix:** 

If possible, interogate ARM to see if registration is really needed up front.

Also, the error message in this case is unhelpful. It just looks like the generic "auth" error from ARM, which is unfortunate because people do struggle with getting the SP working properly and they might not realize the problem. It actually took a couple weeks of sitting in the back of my brain before I realized why this hadn't been working.

cc: @tombuildsstuff

**Comment by colemickens** [comment]
(For the error part, it'd be great if Terraform actually unpacked the inner error and showed it to the user.)

**Comment by tombuildsstuff** [comment]
I had a quick look into this.. [Resource Provider Registration](https://github.com/hashicorp/terraform/blob/master/builtin/providers/azurerm/provider.go#L182) is done through Riviera - so I've [sent a PR](https://github.com/jen20/riviera/pull/22) to Get the Registration State for a particular Resource Provider.

Also from what I can see Terraform is just firing the Registration request and hoping for the best - rather than waiting for the registration to be completed (whilst some Providers register very quickly, some others take a short while - I vaguely recall seeing a race condition here in the past).

Registration is [done in Parallel for the list of Resource Providers](https://github.com/hashicorp/terraform/blob/master/builtin/providers/azurerm/provider.go#L233) - there's almost certainly a performance advantage to pulling in the Registration State for all Providers via the [List endpoint](https://docs.microsoft.com/en-us/rest/api/resources/providers) - and then only registering the Resource Providers which aren't already registered.

Ideally it'd be possible to test if the user has permissions to do this operation and show a helpful message - but I don't see a way of achieving this in the Docs.. potentially a HEAD request might work, but I've not tested that theory (and I don't see it on any of the other Azure API's) - @colemickens are you aware of a means of achieving that?

**Comment by pmcatominey** [comment]
I've got a work in progress fix for this which does the following:

* Lists all providers and their registration status
* Registers only the providers which aren't already registered
* Allows users to opt out of registering any provider in the provider config `skip_provider_registration`

I don't think there will be a big performance difference especially since providers should only have to be registered once per subscription which makes the parallel register requests a bit wasteful. It's worth bearing in mind that not everyone gives their application the permission to register providers as seen in #8824.

The first registration for the Compute provider also serves to detect if the credentials are valid, I've checked whether we can get a better error message back from the API in the case of bad credentials. By using the ProvidersClient#List method with a bad `client_secret`, the error returned by the call yielded the following information:

```
Error refreshing state: 1 error(s) occurred:

* autorest#WithErrorUnlessStatusCode: POST https://login.microsoftonline.com/00000000-0000-0000-0000-00000000/oauth2/token?api-version=1.0 failed with 401 Unauthorized: StatusCode=401
```

Is there an endpoint which will give us a clearer error message @colemickens?
